### DISCUSSION TITLE
[Form] Mapping the values to the object only if the form is valid

### DISCUSSION THREAD

**Discussion by artyuum** [symfony/symfony#49840] [Q&A]
It's possible to declare a field as unmapped, thus the form won't try to map the value to a predefined object and will only keep its value in the form instance. What I like with this approach, is that the constraints are defined directly on the field and there is no risk of breaking the app if there is a type mismatch.

But when declaring a field as mapped, the form will map the value to the object before proceeding to the validation, thus causing an internal error if the type mismatches. 

My questions regarding that are :
1. if it's somehow possible, how can I make the form map the values to the object *only* if the constraints defined on the form fields are respected? I don't care about the constraints defined on the object properties because I keep them in my form.
2. if it's not possible, would a new setting on the form component make sense? The idea would be to validate the form before setting the values to the object. Of course, when this option is used, the constraints would need to be set on the form fields but I believe that this is a common use case considering the fact that a lot of users are facing type mismatch when mapping an entity directly to the form.

**Comment by MatTheCat** [comment]
Try linking dedicated objects to your forms: https://verraes.net/2013/04/decoupling-symfony2-forms-from-entities/ (article is old but still relevant).

**Reply by artyuum** [reply]
This is not really what I want. On top of that, not mapping a form to an entity causes a lot of extra work, mainly when we have to work with collection of entities. It's counter-productive knowing that the Form component handles that perfectly.

I would like to make the Form component map the data **only** if the form is valid (constraints defined on the fields are respected).

**Reply by stof** [reply]
The validator component is used to validate the underlying object, which is why it runs after binding the data.

**Reply by artyuum** [reply]
Here is how I understand the form component. (please, correct me if I'm wrong) 

Once the data is submitted, it's mapped to the object associated with the form (the object defined using the `data_class` option). Let's say I have some unmapped fields in my form with some constraints defined (using the `constraints` option). Their data is kept in the form instance and it's possible to access it with: `$form->get('field')->getData()`. But the validation is still made because it reads the constraints defined on the field side.

What I'm trying to point out (if I understand correctly) is that the form component is **capable** of keeping the values in the form instance and still proceed to the validation, without even mapping to the object defined using the `data_class` (which is not a required option btw), as it's working with unmapped fields.

**So my question is:**
Would it be technically possible to change the behavior of the form component (by making a PR) using an option for example, to force the validation to be made by reading the constraints defined using the `constraints` option, on the field's side, before mapping the data to the final object? (defined using the `data_class` option) That way, we could get rid of this annoying problem when a strongly typed object is mapped to the form, but the form fails to call the setter when passing invalid data. It also prevents the entity from being in an invalid state (if the object is an entity).

I hope I've defined my needs better. Feel free to ask more questions.

**Comment by artyuum** [comment]
https://github.com/symfony/symfony/issues/51579
